sdk/core/Azure.Core/src/Response.cs

```diff
@@ -154,5 +138,150 @@ internal static void DisposeStreamIfNotBuffered(ref Stream? stream)
                 stream = null;
             }
         }
+
+        /// <summary>
+        /// TBD.
+        /// </summary>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        /// <exception cref="NotImplementedException"></exception>
+        public override BinaryData ReadContent(CancellationToken cancellationToken = default)
+        {
+            // Derived types should provide an implementation that allows caching
+            // to improve performance.
+            if (ContentStream is null)
+            {
+                return s_EmptyBinaryData;
+            }
+
+            if (ContentStream is MemoryStream memoryStream)
+            {
+                return memoryStream.TryGetBuffer(out ArraySegment<byte> segment) ?
+                    new BinaryData(segment.AsMemory()) :
+                    new BinaryData(memoryStream.ToArray());
+            }
+
+            BufferedContentStream bufferStream = new();
+
+            Stream? contentStream = ContentStream;
+            contentStream.CopyTo(bufferStream, cancellationToken);
+            contentStream.Dispose();
+
+            bufferStream.Position = 0;
+            ContentStream = bufferStream;
+
+            BinaryData content = BinaryData.FromStream(bufferStream);
+            bufferStream.Position = 0;
+
+            return content;
+        }
+
+        /// <summary>
+        /// TBD.
+        /// </summary>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        /// <exception cref="NotImplementedException"></exception>
+        public override async ValueTask<BinaryData> ReadContentAsync(CancellationToken cancellationToken = default)
+        {
+            // Derived types should provide an implementation that allows caching
+            // to improve performance.
+            if (ContentStream is null)
+            {
+                return s_EmptyBinaryData;
+            }
+
+            if (ContentStream is MemoryStream memoryStream)
+            {
+                return memoryStream.TryGetBuffer(out ArraySegment<byte> segment) ?
+                    new BinaryData(segment.AsMemory()) :
+                    new BinaryData(memoryStream.ToArray());
+            }
+
+            BufferedContentStream bufferStream = new();
+
+            Stream? contentStream = ContentStream;
+            await contentStream.CopyToAsync(bufferStream, cancellationToken).ConfigureAwait(false);
+            contentStream.Dispose();
+
+            bufferStream.Position = 0;
+            ContentStream = bufferStream;
+
+            BinaryData content = BinaryData.FromStream(bufferStream);
+            bufferStream.Position = 0;
+
+            return content;
+        }
+
+        private class BufferedContentStream : MemoryStream { }
```

Comment:
Why do we need this class?

Reply:
I'm not sure we do in the current implementation.  I was planning to dust this up a bit when the PR became real, and I can evaluate whether keeping this is useful one way or another, and if it's not, will remove.

Reply:
Added an explanation in https://github.com/Azure/azure-sdk-for-net/pull/42234